equinox/nn/_batch_norm.py

```diff
@@ -138,38 +203,85 @@ def __call__(
         A `NameError` if no `vmap`s are placed around this operation, or if this vmap
         does not have a matching `axis_name`.
         """
-
         if inference is None:
             inference = self.inference
-        if inference:
-            running_mean, running_var = state.get(self.state_index)
+
+        def _stats(y):
+            mean = jnp.mean(y)
+            mean = lax.pmean(mean, self.axis_name)
+            var = jnp.mean((y - mean) * jnp.conj(y - mean))
+            var = lax.pmean(var, self.axis_name)
+            var = jnp.maximum(0.0, var)
+            return mean, var
+
+        if self.mode == "ema":
+            assert (
+                self.ema_first_time_index is not None
+                and self.ema_state_index is not None
+            )
+            if inference:
+                running_mean, running_var = state.get(self.ema_state_index)
+            else:
+                first_time = state.get(self.ema_first_time_index)
+                state = state.set(self.ema_first_time_index, jnp.array(False))
+
+                batch_mean, batch_var = jax.vmap(_stats)(x)
+                running_mean, running_var = state.get(self.ema_state_index)
+                momentum = self.momentum
+                running_mean = (1 - momentum) * batch_mean + momentum * running_mean
+                running_var = (1 - momentum) * batch_var + momentum * running_var
+                # since jnp.array(0) == False
+                running_mean = lax.select(first_time, batch_mean, running_mean)
+                running_var = lax.select(first_time, batch_var, running_var)
+                state = state.set(self.ema_state_index, (running_mean, running_var))
+
+            def _norm(y, m, v, w, b):
```

Comment:
Am I misreading anything, or are the two copies of this function the same, so that it can be factored out above the `if` statement?